### DISCUSSION TITLE
Max node height

### DISCUSSION THREAD

**Discussion by euu2021** [freeplane/freeplane#1293] [Ideas]
This a follow-up for the discussion that started at  https://github.com/freeplane/freeplane/issues/1279#issuecomment-1616507824

Similar thread [on the old forum](https://sourceforge.net/p/freeplane/discussion/758437/thread/e9200db6c3/)

[Discussion that originated](https://sourceforge.net/p/freeplane/discussion/758437/thread/96d35fd4/) the "minimize node" feature, in 2010. It's an interesting discussion about the usefulness of a "rollup" option

---
> Alternative approach could be implementing `Maximum shortened text width` setting at node _style_ level with default value (used in "Default" node style) taken from global preferences (`Maximum shortened text width`).
> 
> This would be a more flexible solution.

I like to see how your idea evolved from placing the setting in the global Preferences, to placing it into the on individual node settings. This allows many more possibilities, indeed.

I like the idea of, when implementing a new feature, mimic how it is implemented on other places, where it is already tried and tested, and is of common knowledge by the users.
 I see that this feature that we are talking is similar to the "wrap, clip, or overflow" option that is very well-known in spreadsheets. I think that the first thing that any user learns about spreadsheets is that he needs to make that choice ("why is the text of the cell going over the other cell?" 🤣): "wrap, clip, or overflow". 

It's funny that, this week, I was creating a spreadsheet in Notion, and, on the first line, I already went to search how to limit the cell to one line:

![image](https://github.com/freeplane/freeplane/assets/77707706/381d12dd-a295-41f7-9dc3-995620158dbc)

So, I think an interesting way to start thinking about the text presentation on nodes could be with this parallel with "wrap, clip, or overflow" in mind.

PS: feature request at #1292

**Comment by amariusz** [comment]
Overflow is probably not an option in mindmap. It would look messy.
Wrap is current behaviour.
Clip would be `Maximum shortened text width` but in spreadsheets you can set it differently at every column by specifying column width. In Freeplane there's a single setting.

**Comment by amariusz** [comment]
I'll go again with a more concrete answer to make sure we have the same concept in mind. I would like to propose following behavior based on your parallel:

- there's `Maximum shortened text width` setting available in node style (`Core text` section)
- there's a choice (dropdown) between `wrap` / `clip` / `overflow` (also `Core text` section)
- these options affect only minimized nodes (not used otherwise)
- `wrap` - core text is shortened to `Maximum shortened text width` (or first newline) and wrapped. This is current behavior.
- `clip` - core text is shortened to `Maximum shortened text width` (or first newline) AND also clipped after reaching `max node with`. `Maximum shortened text width` could be set to high enough value not to interfere.
- `overflow` - core text is shortened to `Maximum shortened text width` (or first newline) and overflows node edge boundary.

**Reply by euu2021** [reply]
**_About overflow_**: I never understood why Google chooses Overflow as the default behavior in Google Sheets. Now, searching about it, I see that it's more elegant than I initially thought, and is very similar to your idea of having the leaf nodes automatically wider than other nodes.
In Google Sheets, the overflow only happens if there is no content on the cell at the right. In FP, it would be:
-if the node has a child: stay narrow
-if there is no child: go wider


**_About the proposal_**: something that bothers me about the `Maximum shortened text width` is that the size of the text can be very different according to the contents. For example, both those 2 nodes are 29 characters long:
![image](https://github.com/freeplane/freeplane/assets/77707706/ac1de1af-8145-4c9c-8060-54be5376f884)

So, in order to always achieve one liner minimized nodes, I need to choose a low Maximum shortened text width, which usually results in nodes that don't make complete use of the width of the node.
This proposal solves this problem with the Clip option. So, I appreciate it.


**_A new proposal_**: now, I don't understand why to limit the "wrap, clip, or overflow" options to minimized nodes.
I think they can be applied also to "unminimized" nodes, and there is no need to mess with `Maximum shortened text width`:
-wrap: this is the classic FP behavior. So, if max width is reached, the text is wrapped.
-clip: nodes are always one-liners. So, if max width is reached, the text is clipped.
-overflow: as described above: No child? The text goes on.

**Comment by amariusz** [comment]
> **_A new proposal_**: now, I don't understand why to limit the "wrap, clip, or overflow" options to minimized nodes.
> I think they can be applied also to "unminimized" nodes, and there is no need to mess with `Maximum shortened text width`:
> -wrap: this is the classic FP behavior. So, if max width is reached, the text is wrapped.
> -clip: nodes are always one-liners. So, if max width is reached, the text is clipped.
> -overflow: as described above: No child? The text goes on.

Let's assume you have a node with several lines of text. It's "unminimized". Based on the above:
- clip - only part of the text is shown, probably part of first line
- overflow - only part of the text is show - first line that overflows node boundary

For me this is very much like 'minimized' behavior and that's why I suggested limiting those options to minimized nodes.

But your approach is also fine with me.
Correct me if I misunderstood - those options could be in effect all the time but when node is minimized `Maximum shortened text width` is also applied?

**Comment by amariusz** [comment]
> -clip: nodes are always one-liners. So, if max width is reached, the text is clipped.
> -overflow: as described above: No child? The text goes on.

Also - this is not really consistent with how spreadsheets work when there are multiple lines:

![230704_181403_google_sheet_overflow_clip](https://github.com/freeplane/freeplane/assets/20703230/00b8d80e-7c56-46ea-aaab-8a3ab2f1e412)

**Reply by euu2021** [reply]
Yes, good point. I had in mind the Notion behavior, that I've been using lately: if wrap is off, only the first line appears, no matter the number of lines or line breaks in the cell.

As we are in an idea generation stage, we can aggregate all those possibilities. So, I think there are 5 variations, now.

> Correct me if I misunderstood - those options could be in effect all the time but when node is minimized Maximum shortened text width is also applied?

Yes, I think that, anyway, this behavior needs to be kept, because it's already incorporated in the users workflow. I think the ideal is to add more options, making the program more flexible, without suppressing other options.

**Comment by EdoFro** [comment]
> Overflow is probably not an option in mindmap. It would look messy.

I'm very interested in having the "Overflow" format, if posible.

I used the new nodes layouts to [simulate a Gantt Chart](https://github.com/freeplane/freeplane/discussions/1121#discussion-5153758) and in those cases the width of the "tasknode" depends on its duration and not on its text. There the "Overflow" format would be excellent.

> In Google Sheets, the overflow only happens if there is no content on the cell at the right. In FP, it would be:
> -if the node has a child: stay narrow
> -if there is no child: go wider

I think this its not so easy to define:
I would say if a node **has no child OR it's folded**, then it can overflow.
But currently, with some of the nodes layouts you don't have that restriction anymore:
For example when using this: ![image](https://github.com/freeplane/freeplane/assets/35700575/dac79992-acdf-4be7-835c-a777e103c4f5), it produces branches like:

![image](https://github.com/freeplane/freeplane/assets/35700575/f1a57cbd-6310-4aa4-abdf-e412bee5b20f)

where overflowing is not so important.

**Reply by euu2021** [reply]
> I think this its not so easy to define:
I would say if a node has no child OR it's folded, then it can overflow.

Good point. I forgot the folded situation.
